### ISSUE TITLE
real transform support

### ISSUE THREAD

**Issue by drwells** [pyFFTW/pyFFTW#39]
I am interested in wrapping the real-to-real (at least the discrete sine and discrete cosine) transforms. I set up a branch on my fork of pyFFTW and have written an implementation that appears to work. It still needs tests.

My main concern is this: since `FFTW_FORWARD` is an integer, but the real flags (like `FFTW_ROD00`) are enums, the 'correct' way to handle both with the `self._directions` attribute is a union pointer.

It is worth noting that this contradicts the work proposed in #35.

**Comment by JoyMonteiro** [comment]
Hello, I looked for the code with r2r, but it does not seem to be present anywhere. I looked at drwells'
fork, and could not find them either. Any place I can find it. We are planning to use it in a pseudospectral code, so the dct will get tested quite thoroughly ;)

TIA,
Joy

**Comment by drwells** [comment]
It is available under my `r2r` branch.

Pseudospectral methods were my original motivation to write these bindings (although the pseudospectral code is proprietary, alas).

**Comment by JoyMonteiro** [comment]
Thanks! will check them out.

**Comment by rabernat** [comment]
Sorry for resurrecting this old issue, but I thought it would be better than starting a new, redundant one.

Thanks so much for this awesome package! We have built a psuedo-spectral geostrophic turbulence model which uses pyfftw (version 0.9.2).
https://github.com/pyqg/pyqg

We would like to use discrete sine and cosine transforms (DST, DCT) in order to accommodate different boundary conditions. These transforms are supported by FFTW but not pyfftw.

Will the modifications made by @drwells ever be merged into pyfftw? And am I correct that those modifications include DST and DCT support?

**Comment by drwells** [comment]
I don't mind picking this up again (though I am more interested in the MPI features of FFTW at the moment).

Off topic: I am amused that you are using this for the QGEs (thats what I did too). Here, have a movie I made about a year ago:

http://www.math.vt.edu/people/drwells/qge2.webm

A large amount of the code I used to generate that is proprietary, but not all of it is (I have a pretty sophisticated time integrator). I should take a look into what you have done.

**Comment by rabernat** [comment]
Thanks for your reply! If pyfftw doesn't end up supporting DCT, we will probably end up eliminating the dependency and wrapping FFTW directly from within our project. I would prefer to build on existing work obviously, hence my comment above.

Our goal is to build a general purpose python qg model for both teaching and research purposes. It is well documented and stable now, but still lacking lots of features.
http://pyqg.readthedocs.org/

There are lots of these models floating around, and a lot of wheel-reinventing going on. We would love to have your help on pyqg if you are interested in getting involved. Our time integration is quite primitive at this point--Adams Bashforth 3rd order with a fixed timestep. A contribution in this area would be most welcome.

p.s.
Your movie is really cool. Here is one from pyqg:
https://vimeo.com/137749487

**Comment by hgomersall** [comment]
Hi @rabernat,

You're quite right to resurrect it, it is sorely needed. My understanding is we can make a release then roll in @drwells's work. Releasing to pypi for Windows is a real hassle and I just need to commit the hour or two it will take.

I suggest that building on the r2r stuff that @drwells has done is a sensible approach. Any additional changes are likely to be small and would be simple to tweak as and when it is rolled into master. The discussion here would be to have wrapping classes for each specific version, but having the FFTW class as the core, so by using that not much should change (the changes might be things like changing the keyword argument name).

**Comment by rabernat** [comment]
Thanks @hgomersall! I obviously support what you say 100%.

One thing that has really helped our project is to use travis-CI (or appveyor for Windows) to automate testing on different platforms. If you got that going, making incremental releases might not be so painful.

However, it might not work, since your package is so dependent on FFTW libraries.

**Comment by hgomersall** [comment]
Travis is used already, but I've not heard of appveyor until now. It sounds like the solution! @drwells had opened #66 to finalise the 1.0 release.

**Comment by hgomersall** [comment]
@rabernat @drwells I'm thinking we just need to crack on with this rather than waiting for a 1.0 that is not happening. What's the interface that we want for this?

One strategy is to move to an enumerated `transform_type` argument, that encapsulates all the various transforms (including forwards and backwards). We could then deprecate the `direction` argument. Would this work? Do we still need a notion of a direction?

**Comment by hgomersall** [comment]
My previous suggestion is slightly complicated by something that has already been touched on that FFTW takes a per-axis kind argument for r2r transforms. This is logically quite different to the c2r or the r2c transforms. Is it something we want to support?

**Comment by rabernat** [comment]
> One strategy is to move to an enumerated transform_type argument, that encapsulates all the various transforms (including forwards and backwards).

Could you enumerate what you envision as the different options for ``transform_type``?

**Comment by hgomersall** [comment]
So something like:

`transform_obj = pyfftw.FFTW(inp_array, out_array, axes=(2, 3), transform_type=pyfftw.FFTW_REDFT11)`

**Comment by hgomersall** [comment]
or

     transform_obj = pyfftw.FFTW(inp_array, out_array, axes=(2, 3), transform_type=pyfftw.fft_forwards)`

     transform_obj = pyfftw.FFTW(inp_array, out_array, axes=(2, 3), transform_type=pyfftw.dct_IV)`

**Comment by rabernat** [comment]
So the different options enumerated in the [FFTW docs](http://www.fftw.org/doc/Real_002dto_002dReal-Transform-Kinds.html#Real_002dto_002dReal-Transform-Kinds)? That seems reasonable. I prefer the ``dct_IV`` syntax because it is more transparent.

In a multi-dimensional transform, it possible to apply e.g. dct along one axis and regular dft along a different axis?

**Comment by rabernat** [comment]
This seems like a more useful document:
http://www.fftw.org/doc/Real-even_002fodd-DFTs-_0028cosine_002fsine-transforms_0029.html#Real-even_002fodd-DFTs-_0028cosine_002fsine-transforms_0029

**Comment by hgomersall** [comment]
@rabernat No, there is a specific [guru transform](http://www.fftw.org/doc/Guru-Real_002dto_002dreal-Transforms.html#Guru-Real_002dto_002dreal-Transforms) for r2r transforms and a [different one](http://www.fftw.org/doc/Guru-Complex-DFTs.html#Guru-Complex-DFTs) for complex (and also for r2c and c2r). This means we can't generalise to a tuple of tranform types (without implementing it within FFTW).

**Comment by rabernat** [comment]
Let me specify what my application needs. Maybe that will help clarify the optimal api.

We have a field ``p(x,y)`` on the interval ``[(0,Lx), (0,Ly)]``which needs to be Fourier transformed in x and y. Currently we use DFT, which implies that p is doubly periodic.

We would like to support "sidewalls" at either / both x=0,L and y=0,L, on which p = 0. The four cases would then be:

1. doubly periodic: use r2c DFT
2. p(0,y) = 0: use r2r DCT in x direction, r2c DFT in y direction
3. p(y,0) = 0: use r2r DCT in y direction, r2c DFT in x direction
4. p(0,y) = 0 AND p(x,0) = 0: use r2r DCT in both directions

If neither FFTW nor pyfftw supports 2 and 3 (e.g. via a tuple of transform types), then our application would have to extra logic to deal with those cases. Which would be fine of course but, from my perspective, preferable to deal with upstream. However, implementing this in pyfftw would break the simple correspondence with the FFTW api, so I understand that may not be desirable.

**Comment by hgomersall** [comment]
Perhaps it isn't so hard to generalise the FFTW case.

I think in the first instance, we should just support a single transform kind per FFTW object, which is the simplest implementation, and then work out what needs to be done to extend it.

@drwells what's the state of your r2r branch? Is it implemented with tests?

**Comment by drwells** [comment]
Yes (I even got a patch a while back to fix some of the tests). I must confess that I have not had much time to look into this since I am not using these transforms at the moment.

I'll try to clean up what I have an (re)open the PR in a couple of days.

**Comment by hgomersall** [comment]
@drwells that would be great and much appreciated. We can then tune it to tweak the interface. Perhaps the route is to merge into a new branch.

**Comment by rabernat** [comment]
Hi @drwells, just checking in on this.

If there is anything we can do to help with this issue, just let me know. I have some students who might be willing to work on it.

**Comment by drwells** [comment]
Sorry about the delays. I know how to fix this, but I just have not had time (fortunately my schedule is clearing up). I will go back and clean up the PR on Saturday.

I appreciate the offer. I will let you know if I need help.
